### PR TITLE
Request for Quotation

### LINKED ISSUE #307: [New Model]: RequestForQuotation
## Model Description
Request for Quotation model is an information exchange between the buyers and suppliers. Suppliers want to receive all the relevant product manufacturing information and as precisely as possible. This is the role of the RFQ, It includes all the configurations such as material, tolerance class, BoM (bill of material)  and so on. 
This information which is referred to as product manufacturing informations(PMIs) will b forwarded to the suppliers which the buyers will select after the process of matchmaking is done. 

## Relevant Standards

Need to check which of the BoP (Bill of Process) standard can also be referenced with our standard here.

## Example Data

| UseCase   | "Level" | Property                      | mand./opt. | Semantic ID | Default value |
|-----------|---------|-------------------------------|------------|-------------|---------------|
| Function1 |         |                               |            |             |               |
|           |         | requestForQuotationSource     | optional   | string      |               |
|           |         | requestForQuotationId         | mandatory  | string      |               |
|           |         | requestForQuotationName       | mandatory  | string      |               |
|           |         | requestForQuotationDate       | mandatory  | string      |               |
|           |         | materialObject            | mandatory  | matrix?     |               |
|           |         | quantity                      | mandatory  | float       |               |
|           |         | quantityUnit                  | optional   | Enum        |               |
|           |         | generalTolerance              | mandatory  | Enum        |               |
|           |         | requiredDeliveryDate          | mandatory  | string      |               |
|           |         | surfaceTreatment              | optional   | enum        |               |
|           |         | additionalQualityRequirements | optional   | enum        |               |
|           |         | StandardsOffered              |            |             |               |
|           |         | testPartRequired              | optional   | boolean     | false         |
|           |         | additonalComments             | optional   | String      |               |
|           |         | billOfProcess                 | optional   | String      |               |
|           |         | stepFile                      | Mandatory  | String      |               |
|           |         | technischeZeichnung           | Optional   | String      |               |
|           |         | surfaceroughnessfinish        | optional   | float       |               |
|           |         | AnzahlGewinde                | optional   | integer     |               |
|           |         | AnzahlPassungen              | optional   | integer     |               |
|           |         | Warmbehandlung                | optional   | enum        |               |
|           |         | Name                          | Mandatory  | string      |               |
|           |         | Address                       | Mandatory  | string      |               |
|           |         | companyName                   | Optional   | string      |               |
|           |         | emailID                       | Mandatory  | string      |               |
|           |         | rechnungAddress               | Mandatory  | string      |               |
|           |         | deliveryAddress               | Mandatory  | string      |               |
|           |         | phoneNumber                   | Mandatory  | integer     |               |
|           |         | partdescription               | Optional   | string      |               |
|           |         | surfacefinishcolour           | Optional   | enum        |               |
|           |         | 3dTechnology                  | Optional   | enum        |               |
|           |         | technicalDrawingNumber        | Optional   | float       |               |
|           |         | revisionNumber                | Optional   | float       |               |
|           |         | deliveryRequirements          | Optional   | string      |               |
|           |         |                               |            |             |               |
|           |         |                               |            |             |               |

- [x] The proposed aspect model does not exist already
- [x] The proposed aspect model does not extend an existing aspect but introduces completely new functionality
- [x] The proposal references a Catena-X use case
- [x] Relevant standards are mentioned/linked
- [x] A first draft of the data structure is provided 
- [ ] 
[RFQ_Daten final review.xlsx](https://github.com/eclipse-tractusx/sldt-semantic-models/files/12646659/RFQ_Daten.final.review.xlsx)
in an example
[RFQ Example.pdf](https://github.com/eclipse-tractusx/sldt-semantic-models/files/12657270/RFQ.Example.pdf)

Structure : https://confluence.catena-x.net/display/ARTIV/CX+-+0000+Request+for+Quotation+Aspect+Model+v.1.0.0

In case of the 'BoM' data field, the actual structure looks like in the above pasted pdf file. Please support in getting the format right. Appreciate your inputs.

### PR DESCRIPTION
**Pull request by dominikoeh**
*Resiliency Domain/Manufacturing as a Service*

## Description

Implements the new model "Request for Quotation"

Closes #307

## Sample Data

```json
{
  "rfqConfiguration": {
    "firstDeliveryDate": "2023-10-24",
    "additionalFiles": {},
    "cadFile": {},
    "additionalComments": "this is a prototype, recommendations towards design for manufacturing are highly welcome",
    "parts": {
      "generalTolerance": "ISO 2768-1 (m), Rough DIN ISO 2768 cL, Medium DIN ISO 2768 mK, Fine DIN ISO 2768 fH",
      "additionalQualityRequirements": "premium quality check, standard quality check",
      "manufacturingDomain": "additive manufacturing",
      "postProcesses": {
        "postProcessGroup": "heat treatment, surface treatment, surface finishing...",
        "postProcessProperty": "black (for colouring)",
        "postProcessType": "stress relieving, sandblasting, colouring"
      },
      "testPartRequired": false,
      "quantity": {
        "quantityNumber": 2.5,
        "measurementUnit": "unit:litre"
      },
      "material": {
        "materialItems": {
          "type": "density g/cm3",
          "value": "2.7"
        },
        "materialGroup": "aluminum, steel, ..."
      },
      "partId": "Drehteil",
      "billOfProcess": {},
      "manufacturingMethod": "milling, powder bed fusion, welding,...",
      "partName": "Drehteil"
    },
    "billOfMaterial": {
      "catenaXId": "urn:uuid:055c1128-0375-47c8-98de-7cf802c3241d",
      "childItems": [
        {
          "catenaXId": "urn:uuid:055c1128-0375-47c8-98de-7cf802c3241d",
          "quantity": {
            "quantityNumber": 2.5,
            "measurementUnit": "unit:litre"
          },
          "createdOn": "2022-02-03T14:48:54.709Z",
          "businessPartner": "BPNL50096894aNXY",
          "lastModifiedOn": "2022-02-03T14:48:54.709Z"
        }
      ]
    },
    "cadFileType": "STEP, stl, jt",
    "lastDeliveryDate": "2023-12-24"
  },
  "rfqIdentification": {
    "name": "Drehteil",
    "dateTime": "2023-10-24T14:48:54.709Z",
    "source": "https://maasportal.mendixcloud.com/",
    "id": "Drehteil_02_0815"
  },
  "cxHeader": {
    "context": "manufacturing as a service",
    "senderBpn": "BPNL50096894aNXY",
    "messageId": "1DBcedc9-eaC6-5337-C9CC-beCF2d9D8B74",
    "recipientBpn": "BPNL50096894aNXY",
    "sentDateTime": "2023-10-24T14:48:54.709Z",
    "version": "1.0.1"
  },
  "rfqSender": {
    "deliveryRequirements": "no plastic for packaging",
    "senderName": "John Doe",
    "senderAddress": "Sunstreet 1, 5555 Sunstate",
    "senderPhoneNumber": "555 123456",
    "senderEMail": "johndoe@sunny.com",
    "senderDeliveryAddress": "Mystreet 1, 1234 Mystate",
    "senderAccountAddress": "Accountstreet 1, 1234 Accountstate",
    "senderCompanyName": "ManufactureEnterprise"
  }
}
```

## MS2 Criteria
(to be filled out by PR reviewer)
- [x] the model **validates** with the SAMM SDS SDK in the version specified in the Readme.md of this repository by the time of the MS2 check  (e.g., 'java -jar samm-cli.jar aspect \<path-to-aspect-model\> validate ). The  SAMM CLI is available [here](https://eclipse-esmf.github.io/esmf-developer-guide/tooling-guide/samm-cli.html) and in [GitHub](https://github.com/eclipse-esmf/esmf-sdk/releases/tag/v2.3.1)
- [x] use **Camel-Case** (e.g., "MyModelElement" or "TimeDifferenceGmtId", when in doubt follow https://google.github.io/styleguide/javaguide.html#s5.3-camel-case)
- [x] the identifiers for all model elements **start with a capital letter** except for properties
- [x] the identifier for **properties starts with a small letter**
- [x] all model elements **at least contain the fields "preferred name" and "description"** in English language. The description must be comprehensible. It is not required to write full sentences but style should be consistent over the whole model
- [x] Property and the referenced Characteristic should not have the same name
- [x] the versioning in the URN **follows semantic versioning**, where minor version bumps are backwards compatible and major version bumps are not backwards compatible. 
- [x] use **abbreviations only when necessary** and if these are sufficiently common
- [x] **avoid redundant prefixes in property names** (consider adding properties to an enclosing Entity or even adapt the namespace of the model elements, e.g., instead of having two properties `DismantlerId` and `DismantlerName` use an Entity `Dismantler` with the properties `name` and `id` or use a URN like `io.catenax.dismantler:0.0.1`)
- [x] fields `preferredName` and `description` are not the same
- [x] **`preferredName` should be human readable** and follow normal orthography (e.g., no camel case but normal word separation)
- [x] name of aspect is singular except if it only has one property which is a Collection, List or Set. In theses cases, the aspect name is plural.
- [x] units are referenced from the SAMM unit catalog whenever possible
- [x] **use constraints** to make known constraints from the use case explicit in the aspect model 
- [x] when relying on **external standards**, they are referenced through a **"see"** element
- [x] all properties with an [simple type](https://eclipse-esmf.github.io/samm-specification/2.0.0/datatypes.html) have an example value
- [x] metadata.json exists with status "release"
- [x] generated json schema validates against example json payload
- [x] file RELEASE_NOTES.md exists and contains entries for proposed model changes 
- [x] all contributors to this model are mentioned in copyright header of model file

## MS3 Criteria
(to be filled out by semantic modeling team before merge to main-branch)
- [x] All required reviewers have approved this PR (see reviewers section)
- [x] The new aspect (version) will be implemented by at least one data provider
- [x] The new aspect (version) will be consumed by at least one data consumer
- [x] There exists valid test data
- [x] In case of a new (incompatible) major version to an existing version, a migration strategy has been developed
- [x] The model has at least version '1.0.0'
- [x] If a previous model exists, model deprecation has been checked for previous model
- [x] The release date in the Release Note is set to the date of the MS3 approval

### CODE REVIEW

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+
+:lastDeliveryDate a samm:Property ;
+   samm:preferredName "last delivery date"@en ;
+   samm:description "RFQ-Configuration: Date of required and last acceptable delivery"@en ;
+   samm:characteristic :DateCharacteristic .
+
+:firstDeliveryDate a samm:Property ;
+   samm:preferredName "first delivery date"@en ;
+   samm:description "RFQ-Configuration: Date of desired, first, but not must delivery"@en ;
+   samm:characteristic :DateCharacteristic .
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:161]: Please also provide the example value here e.g. '   samm:exampleValue "2023-05-22"^^xsd:date .'
same for `lastDeliveryDate`

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+   samm-c:baseCharacteristic :BpnCharacteristic ;
+   samm-c:constraint :BpnConstraint .
+
+:ContextCharacteristic a samm:Characteristic ;
+   samm:preferredName "context characteristic"@en ;
+   samm:dataType xsd:string .
+
+:CatenaXIdTraitCharacteristic a samm-c:Trait ;
+   samm:description "Trait to ensure UUID v4 data format"@en ;
+   samm-c:baseCharacteristic :Uuidv4Characteristic ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:232]: you can make use of shared aspect model (io.catenax.shared.uuid:1.0.0) for UUID.
you can refer this https://github.com/eclipse-tractusx/sldt-semantic-models/blob/a07bdd5b1515767d4f8d07deb47cf2a4aa728bef/io.catenax.quality_message_content/1.0.0/QualityMessageContent.ttl#L35C9-L35C17

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+:senderBpn a samm:Property ;
+   samm:preferredName "sender BPN"@en ;
+   samm:description "CX-Header: BPN identification of sender partner"@en ;
+   samm:characteristic :BpnTrait ;
+   samm:exampleValue "BPNL50096894aNXY" .
+
+:recipientBpn a samm:Property ;
+   samm:preferredName "recipient BPN"@en ;
+   samm:description "CX-Header: BPN identification of recipient partner"@en ;
+   samm:characteristic :BpnTrait ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:41]: Here also please check if you can make use of shared aspect model https://github.com/eclipse-tractusx/sldt-semantic-models/blob/a07bdd5b1515767d4f8d07deb47cf2a4aa728bef/io.catenax.shared.business_partner_number/1.0.0/BusinessPartnerNumber.ttl

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+
+:context a samm:Property ;
+   samm:preferredName "context "@en ;
+   samm:description "CX-Header: description of use case context"@en ;
+   samm:characteristic :ContextCharacteristic ;
+   samm:exampleValue "manufacturing as a service" .
+
+:messageId a samm:Property ;
+   samm:preferredName "message Id"@en ;
+   samm:description "CX-Header: identitifier used by RFQ protocol"@en ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:52]: typo `identitifier`

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+
+:requestForQuotationId a samm:Property ;
+   samm:preferredName "Id"@en ;
+   samm:description "RFQ-Identification: internal identifier in user account"@en ;
+   samm:characteristic :IdCharacteristic ;
+   samm:exampleValue "Drehteil_02_0815" .
+
+:requestForQuotationName a samm:Property ;
+   samm:preferredName "name"@en ;
+   samm:description "RFQ-Identification: name for the request for user account managment (usually name of the part to be manufactured)"@en ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:80]: typo `managment`

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+   samm:description "RFQ-Configuration: in case there are additional documents or files to share"@en ;
+   samm:characteristic :SetOfFilesCharacteristic .
+
+:senderName a samm:Property ;
+   samm:preferredName "sender name"@en ;
+   samm:description "RFQ-Contact: Name of the sender"@en ;
+   samm:characteristic :SenderNameCharacteristic ;
+   samm:exampleValue "John Doe" .
+
+:senderAdress a samm:Property ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:179]: Adress typo. please check in other places also.

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+:BillOfMaterialCharacteristic a samm:Characteristic ;
+   samm:preferredName "bill of material Characteristic"@en ;
+   samm:dataType :SingleLevelBomAsBuiltReference .
+
+:SetOfManufacturingDomainCharacteristic a samm:Characteristic ;
+   samm:preferredName "set of manufacturing domain characteristic"@en ;
+   samm:dataType :ManufacturingDomain .
+
+:SetOfManufacturingMethodCharacteristic a samm:Characteristic ;
+   samm:preferredName "set of manufacturing mehtod charachteristic"@en ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:279]: typo `charachteristic`

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+:BillOfProcessCharacteristic a samm:Characteristic ;
+   samm:preferredName "bill of process characteristic"@en ;
+   samm:dataType :BillOfProcessReference .
+
+:GeneralToleranceCharacteristic a samm:Characteristic ;
+   samm:preferredName "general tolerance characteristic"@en ;
+   samm:dataType xsd:string .
+
+:AdditionalQualityRequirementsCharacteristic a samm:Characteristic ;
+   samm:preferredName "additional quality requirements charakteristic"@en ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:295]: typo `charakteristic`. please check in other places also.

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+   samm:preferredName "post process property characteristic"@en ;
+   samm:dataType xsd:string .
+
+:NumberOfObjects a samm:Characteristic ;
+   samm:description "Quantifiable number of objects in reference to the measurementUnit"@en ;
+   samm:dataType xsd:double .
+
+:MaterialId a samm:Entity ;
+   samm:preferredName "material Id "@en ;
+   samm:description "the material Id is a detailed element to describe the material and has a value and a type, like value=2.7, type=desity g/m3 or value=AlMgMn, type=chemical "@en ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:540]: do you mean `density`?

**Review by agg3fe — CHANGES_REQUESTED** [review]
please check my comments

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 10]
```diff
@@ -0,0 +1,669 @@
...
+:lastModifiedOn a samm:Property ;
+   samm:preferredName "Last Modification Date"@en ;
+   samm:description "Timestamp when the assembly relationship between parent item and child item was last modified."@en ;
+   samm:characteristic samm-c:Timestamp ;
+   samm:exampleValue "2022-02-03T14:48:54.709Z"^^xsd:dateTime .
+
+:businessPartner a samm:Property ;
+   samm:preferredName "Business Partner"@en ;
+   samm:description "The supplier of the given child item."@en ;
+   samm:characteristic :BpnTrait ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:612]: You can make use of existing characteristic and don't need to define a new one.
For reference: https://github.com/eclipse-tractusx/sldt-semantic-models/pull/405/files#diff-8294717bc9be9d276f548ae9f83e2831503589ef17d4d5bf6f1463cdb397d00eR104

Comment (dominikoeh) [reply io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl]: Thank you for the hint!
I removed the local definition and referenced the `urn:samm:io.catenax.shared.business_partner_number` model

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 11]
```diff
@@ -0,0 +1,654 @@
...
+   samm:dataType xsd:string .
+
+:ContextCharacteristic a samm:Characteristic ;
+   samm:preferredName "context characteristic"@en ;
+   samm:description "description of use case context"@en ;
+   samm:dataType xsd:string .
+
+:CatenaXIdTraitCharacteristic a samm-c:Trait ;
+   samm:description "Trait to ensure UUID v4 data format"@en ;
+   samm-c:baseCharacteristic :Uuidv4Characteristic ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:345]: Here also you can make use of existing characteristic and don't need to define a new one.
For reference: https://github.com/eclipse-tractusx/sldt-semantic-models/pull/405/files#diff-8294717bc9be9d276f548ae9f83e2831503589ef17d4d5bf6f1463cdb397d00eR45

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 12]
```diff
@@ -0,0 +1,654 @@
...
+   samm:preferredName "post process group"@en ;
+   samm:description "process group is depending on the defined manufacturing method and defines the general treatment"@en ;
+   samm:characteristic :PostProcessGroupCharacteristic ;
+   samm:exampleValue "heat treatment, surface treatment, surface finishing..." .
+
+:postProcessType a samm:Property ;
+   samm:preferredName "post process type"@en ;
+   samm:description "specific process out of the process group"@en ;
+   samm:characteristic :PostProcessTypeCharacteristic ;
+   samm:exampleValue "stress releaving, sandblasting, colouring" .
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:551]: typo. use `relieving`

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl [hunk 13]
```diff
@@ -0,0 +1,654 @@
...
+   samm:preferredName "bill of process reference"@en ;
+   samm:description "reference to the bill of process model"@en ;
+   samm:properties ( ) .
+
+:PostProcessData a samm:Entity ;
+   samm:preferredName "post process data"@en ;
+   samm:description "after manufacturing typically post processes are necessary, depending on the manufacturing method"@en ;
+   samm:properties ( :postProcessGroup :postProcessType :postProcessProperty ) .
+
+:Quantity a samm:Entity ;
```
Comment (agg3fe) [inline io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:527]: please add preferredName also here

**Comment by agg3fe** [comment]
@dominikoeh  could you please check my latest comments. then we are good for MS2 afterwards.

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl (hunk 12 above)
Comment (dominikoeh) [reply io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl]: Has been fixed

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl (hunk 13 above)
Comment (dominikoeh) [reply io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl:527]: Has been added

File: io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl (hunk 11 above)
Comment (dominikoeh) [reply io.catenax.request_for_quotation/1.0.0/RequestForQuotation.ttl]: has been added

**Review by agg3fe — APPROVED** [review]

**Pull request merged**